src/Xamarin.Android.Build.Tasks/Utilities/MarshalMethodsNativeAssemblyGenerator.cs

```diff
@@ -260,24 +361,52 @@ void ProcessAndAddMethod (MarshalMethodEntry entry, bool useFullNativeSignature,
 			}
 			Console.WriteLine ();
 
-			methods.Add (method);
-
-			void ThrowInvalidSignature (string signature, string reason)
-			{
-				throw new InvalidOperationException ($"Invalid JNI signature '{signature}': {reason}");
+			if (!overloadedNativeSymbolNames.TryGetValue (method.NativeSymbolName, out List<MarshalMethodInfo> overloadedMethods)) {
+				overloadedMethods = new List<MarshalMethodInfo> ();
+				overloadedNativeSymbolNames.Add (method.NativeSymbolName, overloadedMethods);
 			}
+			overloadedMethods.Add (method);
+
+			allMethods.Add (method);
 		}
 
 		string MangleForJni (string name)
 		{
 			Console.WriteLine ($"    mangling '{name}'");
-			var sb = new StringBuilder (name);
+			var sb = new StringBuilder ();
+
+			foreach (char ch in name) {
+				switch (ch) {
+					case '_':
+						sb.Append ("_1");
+						break;
+
+					case '/':
+						sb.Append ('_');
+						break;
 
-			sb.Replace ("_", "_1");
-			sb.Replace ('/', '_');
-			sb.Replace (";", "_2");
-			sb.Replace ("[", "_3");
-			// TODO: process unicode chars
+					case ';':
+						sb.Append ("_2");
+						break;
+
+					case '[':
+						sb.Append ("_3");
+						break;
+
+					case '$':
+						sb.Append ("_00024");
+						break;
+
+					default:
+						if ((int)ch > 127) {
+							sb.Append ("_0");
+							sb.Append (((int)ch).ToString ("x04"));
+						} else {
+							sb.Append (ch);
```

Comment:
The problem with this is that it'll "pass through" characters which shouldn't be passed through:

  * https://discord.com/channels/732297728826277939/732297837953679412/1009499855674163231
  * https://docs.oracle.com/javase/specs/jvms/se7/html/jvms-4.html#jvms-4.2.2

In particular, Kotlin methods can contain `-`.  (We don't support overriding such methods yet, as we use Java as an intermediary, but if we ever start using Kotlin as an intermediary, we *will* be able to override such methods.)

Thus, what we should instead do is allow *only* `[a-zA-Z0-9]`, and escape *anything else*:

```csharp
if ((ch >= '0' && ch <= '0') ||
        (ch >= 'a' && ch <= 'z') ||
        (ch >= 'A' && ch <= 'Z')) {
   sb.Append (ch);
} else {
    sb.Append ("_0").Append (((int) ch).ToString ("x04"));
}
```

Reply:
I don't know if restricting the allowed character set to `[a-zA-Z0-9]` will yield the same symbol name as computed by JNI. According to [JNI docs](https://docs.oracle.com/javase/8/docs/technotes/guides/jni/spec/design.html#resolving_native_method_names), only non-ASCII characters are encoded in this way. I'd prefer to keep the current algorithm until we actually hit the Kotlin `-` problem.

Reply:
Would make for a fascinating thing to test.  Where's a good Java bytecode assembler?  I need to create a Java method which contains `\n`!